### PR TITLE
feat: connections

### LINKED ISSUE #32: Write concept doc about Connection abstraction & connection upgrade
I think we could stand to clarify what we mean specifically by "connection", and it would help to have an overview of the connection upgrading process.

The upgrading section should link to the secure comms and stream multiplexing docs, since those are both examples of upgraded connections. We should also link to the protocol negotiation (multistream-select) docs, since that's used in the upgrade process.

### PR DESCRIPTION
**Pull request by DannyS03**
## Context

- replaces https://github.com/libp2p/docs/pull/202; also up-to-date with new theme
- adds content about libp2p connection abstractions

## Latest preview

(Outdated preview due to PR not being against master anymore] Please view the latest Fleek preview [here](https://bafybeibqvcs3xgfklk3yxi3b7f3yz33y4okzopldqfka5lpbxsarfirzxi.on.fleek.co/concepts/fundamentals/connections/).

### CODE REVIEW

File: content/concepts/fundamentals/connections.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+The listener's multiaddr determines the security protocol to be used on a connection
+(see [addressing specification]). The multiplexing protocol is determined using protocol
+negotiation, with the multistream-select protocol used for this negotiation process
+(as described in the [Protocol Negotiation section](#protocol-negotiation)). When raw
+connections need both security and multiplexing, security is always established first,
+and the negotiation for stream multiplexing takes place over the encrypted channel.
+
+Here is an example of the connection upgrade process:
+
+- The dialing peer sends a request to initiate a connection to the listening
```
Comment (MarcoPolo) [inline content/concepts/fundamentals/connections.md:115]: This does't seem right. Reference https://github.com/libp2p/specs/blob/master/connections/README.md. The security protocol is negotiated in multistream not in the multiaddr.

File: content/concepts/fundamentals/connections.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+  peer over the underlying transport protocol (e.g., TCP).
+- The listening peer accepts the incoming connection request and sends the security
+  protocol ID (e.g., Noise) in its multiaddr to indicate the security protocol to use.
+- The dialing peer responds with the security handshake message to initiate the Noise
+  protocol.
+- If the security handshake is successful, the peers exchange the multistream protocol
+  ID to establish that they will use multistream-select to negotiate protocols for the
+  connection upgrade.
+- The peers negotiate which stream multiplexer to use by proposing and responding with
```
Comment (MarcoPolo) [inline content/concepts/fundamentals/connections.md:124]: In some cases, the peer may include the stream muxer in the security handshake to save this roundtrip https://github.com/libp2p/specs/blob/master/connections/inlined-muxer-negotiation.md

**Review by MarcoPolo — CHANGES_REQUESTED** [review]

File: content/concepts/fundamentals/connections.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+important to consider compatibility with existing protocols, performance, and security.
+
+### Upgrading connections
+
+libp2p is designed to support a variety of transport protocols, including those that do
+not natively support the core libp2p capabilities of security and stream multiplexing.
+The process of layering capabilities onto "raw" transport connections is known as
+"upgrading" the connection.
+
+The listener's multiaddr determines the security protocol to be used on a connection
```
[the remark this answers is not in the document]
Comment (DannyS03) [reply content/concepts/fundamentals/connections.md]: I jumped the gun here. Updated.

File: content/concepts/fundamentals/connections.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+- Once both nodes have sent and received the close message and acknowledgment, the connection
+  or stream is closed.
+
+### Connection and stream management
+
+libp2p handles the management of connections and streams in a few ways:
+
+- **Tracking**: libp2p keeps track of which connections and streams are open and updates
+  this information as connections and streams are opened or closed.
+- **Timeouts and retries**: If a connection or stream fails, libp2p will automatically retry
```
[the remark this answers is not in the document]
Comment (DannyS03) [reply content/concepts/fundamentals/connections.md]: Hmm, I see what you mean. I modified to [this](https://github.com/libp2p/docs/pull/277/commits/86790d933aae91402b41a0ad0546732cdd14b44f#diff-d63d302b99ec29d312eb0614f7105d1df779526cc8eba743b9b8ad8a2b9c6cc8R55) (I believe this is what can be said).

[the remark this answers is not in the document]
Comment (MarcoPolo) [reply content/concepts/fundamentals/connections.md]: > libp2p only reconnects if you attempt to open a new stream 

I know I said this, but this is only true for go-libp2p. Not how rust-libp2p works. I think js-libp2p is similar to Go in this regard. Either way I think we should remove that line.

**Review by MarcoPolo — APPROVED** [review]
Looks good mod comments.

I think in general when we can point to something in specs we should instead of trying to re-explain it here.

File: content/concepts/introduction/core-abstractions/connections.md [hunk 5]
```diff
@@ -1,7 +1,221 @@
...
+For example, consider a case where two libp2p nodes are attempting to upgrade a
+raw transport connection to a secure libp2p connection using the Noise protocol:
+
+- The dialing peer sends a message containing the protocol ID `/noise/1.0.0` to
+  indicate its desire to use the Noise protocol.
+- The listening peer supports the Noise protocol, which echoes the same protocol ID.
+- Both peers perform the Noise handshake to establish a secure connection.
+
+> If the listening peer did not support the Noise protocol, it would have responded
+> with a different protocol ID or with "na" to indicate that no common protocol
```
[the remark this answers is not in the document]
Comment (DannyS03) [reply content/concepts/introduction/core-abstractions/connections.md]: Yes

**Pull request merged**